sqldb/assets.go

```diff
@@ -0,0 +1,393 @@
+}
+
+func (d *DB) UpdateAsset(ctx context.Context, assetID uint, asset domain.InputAsset) (*domain.Asset, error) {
+	if assetID <= 0 {
+		return nil, errors.New("add id ")
+	}
+	newAsset := &domain.Asset{}
+	switch v := asset.Data.(type) {
+	case *domain.Insight:
+		in := &Insight{}
+		d.db.First(in, assetID)
+		in.FromDomain(v)
+		err := d.db.Save(in).Error
+		if err != nil {
+			return nil, err
+		}
+		newAsset.ID = in.ID
+		newAsset.Data = in.ToDomain()
+	case *domain.Chart:
+		ch := &Chart{}
+		d.db.First(ch, assetID)
+		ch.FromDomain(v)
+		err := d.db.Save(ch).Error
+		if err != nil {
+			return nil, err
+		}
+		newAsset.ID = ch.ID
+		newAsset.Data = ch.ToDomain()
+
+	case *domain.Audience:
+		au := &Audience{}
+		d.db.First(au, assetID)
+		au.FromDomain(v)
+		err := d.db.Save(au).Error
+		if err != nil {
+			return nil, err
+		}
+		newAsset.ID = au.ID
+		newAsset.Data = au.ToDomain()
+	default:
+		return nil, fmt.Errorf("UpdateAsset: %w", ErrThisAssetTypeDoesNotExist)
+	}
+
+	return newAsset, nil
+}
+
+func (d *DB) GetAsset(ctx context.Context, at domain.AssetType, assetID uint) (*domain.Asset, error) {
+	newAsset := &domain.Asset{}
+	switch at {
+	case domain.InsightAssetType:
+		in := &Insight{}
+		err := d.db.First(in, assetID).Error
+		if err != nil {
+			return nil, err
+		}
+		newAsset.ID = in.ID
+		newAsset.Data = in.ToDomain()
+	case domain.ChartAssetType:
+		ch := &Chart{}
+		err := d.db.First(ch, assetID).Error
+		if err != nil {
+			return nil, err
+		}
+		newAsset.ID = ch.ID
+		newAsset.Data = ch.ToDomain()
+	case domain.AudienceAssetType:
+		au := &Audience{}
+		err := d.db.First(au, assetID).Error
+		if err != nil {
+			return nil, err
+		}
+		newAsset.ID = au.ID
+		newAsset.Data = au.ToDomain()
+	default:
+		return nil, fmt.Errorf("GetAsset: %w", ErrThisAssetTypeDoesNotExist)
+	}
+	return newAsset, nil
+}
+
+func (d *DB) DeleteAsset(ctx context.Context, at domain.AssetType, assetID uint) error {
+	switch at {
+	case domain.InsightAssetType:
+		err := d.db.Unscoped().Delete(&Insight{}, assetID).Error
+		if err != nil {
+			return err
+		}
+
+	case domain.ChartAssetType:
+		err := d.db.Unscoped().Delete(&Chart{}, assetID).Error
+		if err != nil {
+			return err
+		}
+
+	case domain.AudienceAssetType:
+		err := d.db.Unscoped().Delete(&Audience{}, assetID).Error
+		if err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("DeleteAsset: %w", ErrThisAssetTypeDoesNotExist)
+	}
+	return nil
+}
+
+func (d *DB) ListAssets(ctx context.Context, query domain.QueryAssets) (*domain.ListedAssets, error) {
+	gormQuery := d.db
+	if query.IsDesc {
+		gormQuery = gormQuery.Where("id < ?", query.LastID).Order("id desc")
+	} else {
+		gormQuery = gormQuery.Where("id > ?", query.LastID)
+	}
+	assets := []domain.Asset{}
+	switch query.Type {
+	case domain.InsightAssetType:
+		ins := []Insight{}
+		err := gormQuery.Limit(query.Limit).Find(&ins).Error
+		if err != nil {
+			return nil, err
+		}
+		assets = listRowsToAssets(ins)
```

Comment:
How would you change the listing part of code in case you have millions of entities?

No need for working code, just a list of changes necessary with some hints as to why.

Reply:
First, I would remove the listRowsToAssets by removing the domain.Asset wrapper and refactoring the code to use the domain.AssetType for recognizing the type of the asset. 
The reason is that listRowsToAssets have O(n) debt and finding the asset type through switch v := asset.Data.(type) is slower.
Moreover, I would replace the Find method with FindInBatches and the Create method with CreateInBatches, as it is recommended from their "Performance" section of their documentation.